server/src/main/java/org/elasticsearch/common/util/BytesRefHash.java

```diff
@@ -34,13 +40,31 @@ public BytesRefHash(long capacity, BigArrays bigArrays) {
 
     // Constructor with configurable capacity and load factor.
     public BytesRefHash(long capacity, float maxLoadFactor, BigArrays bigArrays) {
+        this(null, capacity, maxLoadFactor, bigArrays);
+    }
+
+    public BytesRefHash(BytesRefArray bytesRefArray, BigArrays bigArrays) {
+        this(bytesRefArray, DEFAULT_MAX_LOAD_FACTOR, bigArrays);
+    }
+
+    public BytesRefHash(BytesRefArray bytesRefs, float maxLoadFactor, BigArrays bigArrays) {
+        this(bytesRefs, bytesRefs.size() + 1, maxLoadFactor, bigArrays);
+
+        // recreate hashes
+        for (int i = 0; i < bytesRefs.size(); ++i) {
+            bytesRefs.get(i, spare);
+            reset(rehash(spare.hashCode()), i);
+        }
+    }
+
+    private BytesRefHash(BytesRefArray byteRefs, long capacity, float maxLoadFactor, BigArrays bigArrays) {
```

Comment:
I thought of a hypothetical situation, but I'm not sure if it's an issue or not. In the case where we deserialize the ByteRefArray and pass it in to the constructor of BytesRefHash, we will sometimes release this reference and sometimes not, depending on if super fails to allocate and/or throws an exception. 

The super call isn't in the try catch, so if the super fails we will not call release on the passed in BytesRefArray, assuming the caller of the constructor would be responsible for that. However, if we fail to allocate the hashes, we've already set the reference `this.bytesRefs`, so the close() call would release BytesRefArray. 

My question is, do we ever want to release a passed in BytesRefArray if we haven't allocated it here? Perhaps there's no danger in releasing a reference twice, however releasing it sometimes might cause bugs in the caller code if they don't expect the BytesRefArray to be released by being passed in.

Swapping the lines such that hashes is allocated first would eliminate the problem, but I'm still not sure how do we document or inform users of the passed in byteRefs that releasing is delegated to the ByteRefsHash object once passed in.

Reply:
Perhaps writing tests for these exceptional situations would ensure we have the right contracts in place.

Reply:
I documented the correct usage and created 2 tests, see https://github.com/elastic/elasticsearch/pull/85826/commits/f670bd928a23e38304ef0008da53850fd0c544f2